feat: redact option

gjj opened the pull request:
Goals of this MR:

- [x] Sensitive fields will first be marked as ready to be redacted
- [x] Allow users to redact other fields too (using a toggle)

gjj committed:
feat: hooks + download btn

gjj committed:
chore: upgrade oa

gjj committed:
fix: moved getData to DocumentViewer instead

gjj committed:
fix: lint

gjj committed:
fix: obfuscate paths

gjj committed:
feat: state lift

gjj committed:
chore: stuck

gjj commented on src/components/DocumentViewer/DocumentViewer.tsx:
```diff
@@ -49,23 +56,40 @@ export const DocumentViewer: React.FunctionComponent<DocumentViewerProps> = ({ d
+      sensitiveFields.forEach((row: { path: string }) => {
+        _redactionList.push(row.path);
+      });
+      // setRedactionList(_redactionList);
```
this somehow causes infinite re-rendering so I think it shouldn't be placed here, and instead should be placed at PrivacyFilter.tsx? (see https://github.com/Open-Attestation/privacy-filter/pull/18/files#r455585417)

the purpose of this is: on file upload, I'll do one, and only one check for sensitive fields, then push them into the `redactionList`. checking it here seems like it causes an infinite loop...

Nebulis replied:
the reason is simple, when you will call `setRedactionList(_redactionList)`, a state change will be triggered, the component will be rendered but the call to `setRedactionList(_redactionList)` happens during rendering hence the infinite loop

Have a look to `useEffect`

gjj commented on src/components/PrivacyFilter/PrivacyFilter.tsx:
```diff
@@ -41,6 +46,8 @@ export const PrivacyFilter: React.FunctionComponent = () => {
       reader.onerror = () => console.log("File reading has failed.");
       reader.onloadend = () => {
+        // Reset redaction list
+        setRedactionList([]);
```
i was thinking of placing the sensitive fields checker right after this, so that it doesn't cause an infinite loop - is my understanding correct?

gjj committed:
fix: states :worried:

gjj committed:
fix: states

shouldn't need to rely on redactionList here, since it's on init. using redactionList as dep here would cause an infinite loop.. as i'm setting redactionList later on

gjj committed:
feat: styling and recommendations table

gjj committed:
chore: cleanup

Nebulis commented on src/components/DocumentViewer/DocumentViewer.tsx:
```diff
@@ -1,40 +1,46 @@
 
 interface DocumentViewerProps {
-  document?: any;
+  wrappedDocument?: any;
```
there is a type for this => WrappedDocument<any> from open attestation

gjj replied:
thanks for pointing it out! I was using WrappedDocument in PrivacyFilter, forgot about it here

Nebulis commented on src/components/DocumentViewer/DocumentViewer.tsx:
```diff
@@ -1,40 +1,46 @@
+
+  const download = (): void => {
+    const redacted = obfuscateDocument(wrappedDocument, redactionList);
+    console.log(redactionList, redacted);
```
to remove ?

gjj replied:
removed!

Nebulis commented on src/components/PrivacyFilter/PrivacyFilter.tsx:
```diff
@@ -1,13 +1,36 @@
+    setRedactionList(redactionList);
+  };
+
+  const wrappedDocument = document ? document : {};
```
document ?? {}

Nebulis replied:
why is this needed ?

Nebulis commented on src/components/RecommendationsDisplay/RecommendationsDisplay.tsx:
```diff
@@ -1,45 +1,42 @@
 
 interface RecommendationsDisplayProps {
-  hasSensitiveFields?: any;
+  sensitiveFields?: any;
```
string[]  ?

gjj replied:
updated!

gjj committed:
fix: logs, code cleanup

gjj committed:
fix: types
